### PR TITLE
HelpScout integration

### PR DESCRIPTION
**Pull request by zoldar**
### Changes

This PR implements HelpScout integration making easier to asses customer state directly from HS dashboard.

Once reviewed, the migration will be eventually split out from this PR and deployed before the rest.

### Tests
- [x] Automated tests have been added

### Changelog
- [x] This PR does not make a user-facing change

### Documentation
- [x] This change does not need a documentation update

### Dark mode
- [x] This PR does not change the UI

### CODE REVIEW

File: extra/lib/plausible/help_scout.ex [hunk 1]
```diff
@@ -0,0 +1,259 @@
...
+  @signature_field "X-HelpScout-Signature"
+
+  @doc """
+  Validates signature against secret key configured for the
+  HelpScout application.
+
+  NOTE: HelpScout signature generation procedure at
+  https://developer.helpscout.com/apps/guides/signature-validation/
+  fails to mention that it's implicitly dependent on request params
+  order getting preserved. PHP arrays are ordered maps, so they provide
```
Comment (aerosol) [inline extra/lib/plausible/help_scout.ex:25]: tyrranic

File: extra/lib/plausible/help_scout.ex [hunk 2]
```diff
@@ -0,0 +1,259 @@
...
+  def get_customer_details(customer_id) do
+    with {:ok, emails} <- get_customer_emails(customer_id),
+         {:ok, user} <- get_user(emails) do
+      user = Plausible.Users.with_subscription(user.id)
+      plan = Billing.Plans.get_subscription_plan(user.subscription)
+
+      {:ok,
+       %{
+         status_label: status_label(user),
+         status_link: PlausibleWeb.Endpoint.url() <> "/crm/auth/user/#{user.id}",
```
Comment (aerosol) [inline extra/lib/plausible/help_scout.ex:76]:
```elixir
Routes.kaffy_resource_url(PlausibleWeb.Endpoint, :show, :auth, :user, user.id)
```

File: extra/lib/plausible/help_scout.ex [hunk 3]
```diff
@@ -0,0 +1,259 @@
...
+    Billing.PaddleApi.vendors_domain() <>
+      "/subscriptions/customers/manage/" <> paddle_id
+  end
+
+  defp status_label(user) do
+    subscription_active? = Billing.Subscriptions.active?(user.subscription)
+    trial? = Plausible.Users.on_trial?(user)
+
+    cond do
+      not subscription_active? and not trial? and user.trial_expiry_date == nil ->
```
Comment (aerosol) [inline extra/lib/plausible/help_scout.ex:95]:
```suggestion
      not subscription_active? and not trial? and is_nil(user.trial_expiry_date) ->
```

File: extra/lib/plausible/help_scout.ex [hunk 4]
```diff
@@ -0,0 +1,259 @@
...
+        "None"
+
+      user.subscription == nil and not trial? ->
```
Comment (aerosol) [inline extra/lib/plausible/help_scout.ex:98]:
```suggestion
      is_nil(user.subscription) and not trial? ->
```

File: extra/lib/plausible/help_scout.ex [hunk 5]
```diff
@@ -0,0 +1,259 @@
...
+  defp plan_label(_, nil) do
+    "None"
+  end
+
+  defp plan_label(_, :free_10k) do
+    "Free 10k"
+  end
+
+  defp plan_label(subscription, %Billing.Plan{} = plan) do
+    [plan] = Billing.Plans.with_prices([plan], "127.0.0.1")
```
Comment (aerosol) [inline extra/lib/plausible/help_scout.ex:131]: maybe the default ip should be kept at `with_prices` definition. unnecessary confusion here.

File: extra/lib/plausible/help_scout.ex [hunk 6]
```diff
@@ -0,0 +1,259 @@
...
+    interval = Billing.Plans.subscription_interval(subscription)
+    quota = PlausibleWeb.AuthView.subscription_quota(subscription, [])
+
+    price =
+      cond do
+        interval == "monthly" && plan.monthly_cost ->
```
Comment (aerosol) [inline extra/lib/plausible/help_scout.ex:137]: nice, apparently the costs may or may not be there

File: extra/lib/plausible/help_scout.ex [hunk 7]
```diff
@@ -0,0 +1,259 @@
...
+        true ->
+          "N/A"
+      end
+
+    "#{quota} Plan (#{price} #{interval})"
+  end
+
+  defp plan_label(subscription, %Billing.EnterprisePlan{} = plan) do
+    quota = PlausibleWeb.AuthView.subscription_quota(subscription, [])
+    price_amount = Billing.Plans.get_price_for(plan, "127.0.0.1")
```
Comment (aerosol) [inline extra/lib/plausible/help_scout.ex:152]: ditto ip noise

File: extra/lib/plausible/help_scout.ex [hunk 8]
```diff
@@ -0,0 +1,259 @@
...
+    else
+      {:error, :not_found}
+    end
+  end
+
+  defp get_customer_emails(customer_id, opts \\ []) do
+    refresh? = Keyword.get(opts, :refresh?, true)
+    token = get_token!()
+
+    url = @base_api_url <> "/v2/customers/" <> customer_id
```
Comment (aerosol) [inline extra/lib/plausible/help_scout.ex:180]: `Path.join` just to be safe?

File: extra/lib/plausible/help_scout.ex [hunk 9]
```diff
@@ -0,0 +1,259 @@
...
+        if refresh? do
+          refresh_token!()
+          get_customer_emails(customer_id, refresh?: false)
+        else
+          {:error, :auth_failed}
+        end
+
+      error ->
+        Sentry.capture_message("Failed to obtain customer data from HelpScout API",
+          extra: %{error: inspect(error)}
```
Comment (aerosol) [inline extra/lib/plausible/help_scout.ex:205]: let's include customer id just in case?

File: lib/plausible/application.ex [hunk 10]
```diff
@@ -111,6 +113,22 @@ defmodule Plausible.Application do
     :ok
   end
 
+  defp start_help_scout_vault(children) do
+    help_scout_vault_key =
+      :plausible
+      |> Application.fetch_env!(Plausible.HelpScout)
+      |> Keyword.fetch!(:vault_key)
+
+    if ee?() && help_scout_vault_key do
```
Comment (aerosol) [inline lib/plausible/application.ex:122]: consider moving ee check to compile-time?

Something like:

```elixir
on_ee do
  children = children ++ help_scout_vault()
end
```

Reasoning: `vault_key` is always defined in `ee` builds.

Also, should the `HelpScout` module land in `extra/`?

File: lib/plausible/sites.ex [hunk 11]
```diff
@@ -338,6 +338,14 @@ defmodule Plausible.Sites do
     )
   end
 
+  def owned_sites_locked?(user) do
+    user
+    |> owned_sites_query()
+    |> select([site], site.locked)
+    |> Repo.all()
+    |> Enum.any?(& &1)
```
Comment (aerosol) [inline lib/plausible/sites.ex:346]: feels like postgres could do it?

**Review by aerosol — COMMENTED** [review]
Nice work!

File: lib/plausible/application.ex (hunk 10 above)
Comment (zoldar) [reply lib/plausible/application.ex]: https://github.com/plausible/analytics/pull/4327/commits/d38bc0dd9baf0850d2d877c5d9e62dff1668ff94

File: extra/lib/plausible/help_scout.ex (hunk 9 above)
Comment (zoldar) [reply extra/lib/plausible/help_scout.ex]: https://github.com/plausible/analytics/pull/4327/commits/5c113db4eb35c76c6354ba64456d12c298202e8d

File: extra/lib/plausible/help_scout.ex (hunk 2 above)
Comment (zoldar) [reply extra/lib/plausible/help_scout.ex]: https://github.com/plausible/analytics/pull/4327/commits/1a3a87621e368180388a80e81f80ea58a03ede5a

**Review by aerosol — APPROVED** [review]

**Pull request merged**